docs/concepts.md

```diff
@@ -67,6 +68,96 @@ that are specific to a particular Node (like per-process UNIX file descriptors),
 for Node B (unless there happens to be a coincidence of numbering).
 
+## Labels
+
+Labels are used to enforce information flow control (IFC) between nodes and
+channels in an Oak Application and the outside world.
+
+Each node and each channel is associated with a label at creation time. The
+label does not change any more after that (apart from explicit declassification
+operations, which are not currently supported).
+
+A label has two components: **secrecy** and **integrity**.
+
+Each component is represented as an (unordered) set of tags, and each tag refers
+to individual security principal in the system.
+
+The following types of security principals are supported in Oak:
+
+- **user**: specified by a bearer token, public key, or some other assertion of
+  a user identity
+- **computation**: specified by a measurement (hash) of the WebAssembly module
+  running within an Oak Node
+
+The least privileged label is "public_trusted" (also referred to as `⊥`,
+pronounced "bottom"), which corresponds to a node or channel which has only
+observed public data, and its inputs are not endorsed with any level of
+integrity; in this label, both secrecy and integrity components are empty sets.
+
+Given two labels `L_a` and `L_b`, `L_a ⊑ L_b` (pronounced "flows to") iff `L_a`
+is at least as strict as `L_b`, i.e. `b` has "more secrecy" and "less integrity"
+than `a`.
+
+Intuitively, data can only flow to nodes or channels with:
+
+- **at least** that level of secrecy
+- **at most** that level of integrity
+
+For instance, if integrity of `a` is `{s_0, s_1}`, and integrity of `a` is
+`{i_0, i_1}`, then:
+
+- data can flow from `a` to `b` if `b`'s secrecy is `{s_0, s_1, s_2}`, since it
+  corresponds to "more secret" data
+- data cannot flow from `a` to `b` if `b`'s secrecy is `{s_0}`, since it
+  corresponds to "less secret" data
+- data cannot flow from `a` to `b` if `b`'s integrity is `{i_0, i_1, i_2}`,
+  since it corresponds to "more endorsed" data
+- data can flow from `a` to `b` if `b`'s integrity is `{i_0}`, since it
+  corresponds to "less endorsed" data
+
+In particular:
+
+- A node with label `L_w` may write to a channel with label `L_c` iff
+  `L_w ⊑ L_c`.
+- A node with label `L_r` may read from a channel with label `L_c` iff
+  `L_c ⊑ L_r`.
+
+If a node tries to write to or read from a channel that it is not allowed to,
+based on the `⊑` relation, the operation fails immediately, no side effects are
+performed.
+
+It follows that bi-directional communication between nodes `a` and `b` is only
+allowed (via two uni-directional channels) if
+`(L_a ⊑ L_b) ∧ (L_b ⊑ L_a) ⇒ L_a = L_b`, i.e. if `a` and `b` have identical
+secrecy and integrity.
+
+### gRPC and user labels
+
+For incoming gRPC connections, each gRPC invocation is represented b a message
+containing two channels:
+
+- a "request receiver" channel whose label has:
+
+  - a secrecy component explicitly specified by the caller as gRPC request
```

Comment:
I think clients that are external to the system should not be allowed to specify the labels used to handle their own calls because I suspect that it permits hidden declassification and potential attacks. The label used here should be an atomic one determined by the gRPC based exclusively based on the identity of the client machine. The gRPC should then simply hand off the call to an oak node with the label it determines from the client. This node might then also run yet another gRPC (because now it is actually the thing doing remote procedure calls), but now it is constrained by the label of the client, and eventually when they are implemented, the rules for downgrading. With this design, the un-labeled gRPC (which is implicitly completely trusted) has no logic related to labels, so it is a reduction in trust assumptions.

Reply:
Thanks, this makes sense to me, but in that case, what should the label of the gRPC node be? In particular, what is its secrecy component of that node, before downgrading? Can it just be "public_trusted"?

Reply:
If I recall correctly, the gRPC is currently not running in a node and this is the boundary between client apps and the Oak runtime. The gRPC is effectively not labeled, right? In this case it seems useful to think of it as public_trusted (even though it is not really running in an oak node).

If this is actually doable, I think it would be useful to have the main gRPC just look up a fixed label for specific client (and not dependent on what request the client has made). This gRPC can then forward the function call to an actual node that really looks up how to handle a particular request To disambiguate, I'll call the public_trusted (really un-labeled handler) the gRPC and the labeled node responsible for handling requests an LRPC. Maybe the sensible way to do this is to have this LRPC node also run gRPC, though I do not really know enough about that :).

As an example, if Alice does a send_message call, the main gRPC would decide that this request is coming from Alice and just forward this request to an LRPC node that has Alice's label. Since the LRPC has label  ({Alice}, {Alice}) it could downgrade this message to ({}, {Alice}) so that it can be sent to a chat room from which Bob can read.  The LRPC also does the work of figuring out how to actually hand the send_message call whereas the gRPC just transparently passes it over to the LRPC. No matter what request the Alice machine makes, the gRPC always forwards the request to really be handled by an LRPC  with Alice's label. This ensures that requests by Alice cannot cause Alice to downgrade more information than should be allowed.

Given that your system does have integrity labels, it already seems important to ensure that clients cannot specify their own integrity labels. Alice should not be able to specify that she is trusted by Bob when making a call to Oak.

If the the upcoming declassification proposal intends to enforce robustness, letting clients decide their integrity labels will also give them a way to downgrade secrets they do not own.

This gRPC --> LRPC design is roughly what I was thinking after we were talking about the chat app on the whiteboard. I think I have a rough sketch of how labeled transactions could look for this system during the chat app. If this is useful/interesting, I could look at those notes again and we can talk about it in person to see if it makes sense? 

As a side point, for the chat room app, I keep wanting to have disjunctive principals for chat rooms. For example, {Alice \/ Bob } seems like a good secrecy label for a chat room with those two participants. I suppose we can fake this with a third tag called AliceOrBob as discussed before (where AliceOrBob > Alice and AliceOrBob > Bob). So maybe, a more realistic version of the above would be that Alice's LRPC does a downgrading send call to the chat room  from ( {Alice}, {Alice} ) to ( {AliceOrBob},  {Alice} ). To see that this downgrading send is robust we can split it into two steps: an downgrade from ( {Alice}, {Alice} ) to ( {}, {Alice} ) which is robust because we can subtract the integrity tag from the confidentiality set to get there, followed by an upgrade from ( {}, {Alice} ) to ( {AliceOrBob}, {Alice} ) where upgrades (and upgrades are movements upwards in the lattice -- for integrity upgrades go the opposite way :) ) are just always allowed.

Reply:
Yes this makes sense, but I need to think more how to reconcile it with the actual implementation details of the gRPC node in Oak. I'll leave as it is for now, but I'll revisit shortly after.